streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -575,17 +577,22 @@ private void setRepartitionTopicMetadataNumberOfPartitions(final Map<String, Int
                             }
                         }
 
-                        // if we still have not found the right number of partitions,
-                        // another iteration is needed
                         if (numPartitions == null) {
                             numPartitionsNeeded = true;
+                            log.trace("Unable to determine number of partitions for {}, another iteration is needed",
+                                      repartitionSourceTopic);
                         } else {
-                            repartitionTopicMetadata.get(topicName).setNumberOfPartitions(numPartitions);
+                            repartitionTopicMetadata.get(repartitionSourceTopic).setNumberOfPartitions(numPartitions);
+                            progressMadeThisIteration = true;
                         }
                     }
                 }
             }
-        } while (numPartitionsNeeded);
+        } while (numPartitionsNeeded && progressMadeThisIteration);
```

Comment:
Would it be easier to do
```
  if (!progressMadeThisIteration) {
    throw new TaskAssignmentException("Failed to compute number of partitions for all repartition topics");
  }
} while (numPartitionsNeeded);
```
(and don't have the `if (numPartitionsNeeded)` after the while loop)

I guess both is correct and it might be a matter of taste. Also don't have a strong opinion about it. It's just an idea.

Reply:
Mm yeah that does sound cleaner, will do